Omnibus effectively-once bugfixes and test scripts

slfritchie opened the pull request:
This is a big PR, sorry.  The main reviewer will be John.  For y'all else, if you have a chance to look this over, please focus first on source files that you're familiar with and have an opinion about.

- Many bugfixes to improve end-to-end effectively-once data handling
  inside of Wallaroo.  See original `aloc-tee-source` branch history
  for more detail.

    - Fix 2PC error when no connector sink data was written during checkpoint's interval
    - Bug #2999 fixup: app_name + worker_name
    - remove _mute_local() and _unmute_local() + ConnectorSourceCoordinator _start_sources
    - Add reconnect loop to ALO connector, fixes #2722
    - Fix source=None error in MultiSourceConnector
    - ConnectorSink to switch processing phase for many types of barriers
    - add twopc_txn_id_last_committed_helper() function, avoid bogus Fail() by checkpoint after cluster grows
    - workaround for GH #3002
    - wrap checkpoint_initiator_phase.pony Invariant() in ifdef "enforce_worker_invariant"
    - really dispose OutgoingBoundary
    - avoid Fatal() after crash & restart PHASE II + checkpoint_complete()
    - ConnectorSourceNotify: checkpoint: always write at least one byte to EventLog
    - fix relinquish of stream id on rollback, always close socket after sending RESTART
    - major overhaul of correctness testing in aloc_sink (several parts)
    - adjust Fail() criteria for twopc_phase1_reply
    - prevent CheckpointBarrierToken(0) from escaping outside of connector_sink_twopc.pony
    - fix internal bookkeeping for stream position in aloc_sink
    - fix key errors in aloc_sink's _streams dict
    - fix corner case where ConnectorSink must force 2PC abort on rollback
    - stumbled into real race condition in rollback() where 2PC state was commit but we didn't abort, ouch
    - fix race when ConnectorSink reconnects & needs to abort, send NotifyMsg first if we haven't notified yet
    - use correct message_id/file offset in ThrottledFileReader :facepalm:
    - preserve queued messages when switching from BarrierSinkPhase to QueuingSinkPhase

- Small changes to other source & sink classes & their builder
  wrappers to add new state data, e.g., passing the Wallaroo app name
  string to sinks, adding a `BarrierCoordinator` argument to sources.

- Add new `LogCategory` types: `global_stream_registry()` and
  `local_stream_registry`; converted `printf()`-style logging to newer
  `Log`-style.

- Add `aloc_passthrough` app to the `examples/pony` collection.  This
  app uses the `ConnectorSource` and `ConnectorSink` for source & sink
  rather than the not-so-intelligent TCP standard source & sink.

- Add the `testing/correctness/scripts/effectively-once` script
  collection for testing end-to-end effectively-once systems by
  starting, crashing, restarting, and shrink & grow operations.
  The scripts assume that all processes will run on a single machine
  and communicate via the loopback interface via IP address 127.0.0.1.

- Small fixes to the connector source & sink implementation and demo
  scripting, such as changing the default cookie string.

slfritchie commented:
I'll investigate the aloc_sink test failure on Wednesday.

jtfmumm commented on lib/wallaroo/core/source/gen_source/gen_source.pony:
```diff
@@ -122,7 +122,8 @@ actor GenSource[V: Any val] is Source
     source_registry: SourceRegistry,
-    disposable_registry: DisposableRegistry)
+    disposable_registry: DisposableRegistry,
+    barrier_coordinator_: BarrierCoordinator)
```
Is this used?

slfritchie replied:
Removed by commit ea90d7ff97c891fdb5f0c8b06cd238cf49e96b62. I'll mark the other comments related to this arg as resolved, also.

jtfmumm commented on lib/wallaroo/core/source/gen_source/gen_source_coordinator.pony:
```diff
@@ -88,7 +89,8 @@ actor GenSourceCoordinator[In: Any val] is SourceCoordinator
     recovering: Bool, target_router: Router,
-    generator: GenSourceGeneratorBuilder[In])
+    generator: GenSourceGeneratorBuilder[In],
+    barrier_coordinator: BarrierCoordinator)
```
Do we need this threaded along?  It doesn't look like it's used by GenSource

jtfmumm commented on lib/wallaroo/core/source/gen_source/gen_source_coordinator_builder.pony:
```diff
@@ -45,6 +46,7 @@ class val GenSourceCoordinatorBuilder[In: Any val]
   let _auth: AmbientAuth
   let _layout_initializer: LayoutInitializer
   let _recovering: Bool
+  let _barrier_coordinator: BarrierCoordinator
```
Same question, one level up

jtfmumm commented on lib/wallaroo/core/checkpoint/checkpoint_initiator_phase.pony:
```diff
@@ -124,7 +124,9 @@ class _CheckpointingPhase is _CheckpointInitiatorPhase
     ifdef debug then
       Invariant(checkpoint_id == _token.id)
-      Invariant(_c_initiator.workers().contains(worker))
+      ifdef "enforce_worker_invariant" then
```
Why has this been moved into an ifdef?

slfritchie replied:
I would've sworn that that `enforce_worker_invariant ` appeared elsewhere in the code, and that I'd added this one with the same name.  But I can't find another instance.  Hm.

slfritchie replied:
Removed by commit f697a511ad108d8a7b7712640558feb0cb30bb3a

jtfmumm commented on lib/wallaroo/core/sink/kafka_sink/kafka_sink_config.pony:
```diff
@@ -73,7 +73,8 @@ class val KafkaSinkBuilder
     reporter: MetricsReporter iso, env: Env,
     barrier_coordinator: BarrierCoordinator, checkpoint_initiator: CheckpointInitiator,
-    recovering: Bool, worker_name_: WorkerName, auth_: AmbientAuth): Sink
+    recovering: Bool, app_name_: String, worker_name_: WorkerName,
```
Why `app_name_`?  Is this because of shadowing?  In that case, the convention is `app_name'`.  Same for `worker_name_` and `auth_`.

jtfmumm commented on lib/wallaroo/core/sink/tcp_sink/tcp_sink_builder.pony:
```diff
@@ -115,12 +115,16 @@ class val TCPSinkBuilder
     barrier_coordinator: BarrierCoordinator, checkpoint_initiator: CheckpointInitiator,
-    recovering: Bool, worker_name_: WorkerName, auth_: AmbientAuth): Sink
+    recovering: Bool, app_name: String, worker_name: WorkerName,
+    auth_: AmbientAuth): Sink
```
same points regarding naming

slfritchie replied:
I'm trying to indicate that the variable is unused.

slfritchie replied:
The `SinkBuilder` interface requires the last `AmbientAuth` arg. `ConnectorSink` needs it, but `TCPSink` does not.  I'll remove the trailing underscore.

jtfmumm commented on lib/wallaroo/core/sink/tcp_sink/tcp_sink_builder.pony:
```diff
@@ -115,12 +115,16 @@ class val TCPSinkBuilder
       .cstring())
 
-    let id: RoutingId = RoutingIdGenerator()
+    let id: RoutingId = try DeterministicSourceIdGenerator(
```
Wasn't this renamed so that it doesn't refer to `SourceId`?  Something like `RoutingIdFromStringGenerator`

slfritchie replied:
IIRC that change is what's causing the conflict with tcp_sink_builder.pony and perhaps the other conflict, too: the base of this branch is a bit old.

jtfmumm commented on lib/wallaroo/core/source/connector_source/connector_source.pony:
```diff
@@ -117,7 +120,8 @@ actor ConnectorSource[In: Any val] is (Source & TCPActor)
     source_registry: SourceRegistry,
-    disposable_registry: DisposableRegistry)
+    disposable_registry: DisposableRegistry,
+    barrier_coordinator: BarrierCoordinator)
```
This isn't used, is it?

jtfmumm commented on lib/wallaroo/core/source/connector_source/connector_source_coordinator.pony:
```diff
@@ -131,6 +133,7 @@ actor ConnectorSourceCoordinator[In: Any val] is
     handler: FramedSourceHandler[In] val,
     host: String, service: String, cookie: String,
     max_credits: U32, refill_credits: U32,
+    barrier_coordinator: BarrierCoordinator,
```
This is threaded down, but doesn't appear to be used.

jtfmumm commented on lib/wallaroo/core/source/connector_source/connector_source_coordinator_builder.pony:
```diff
@@ -70,7 +72,8 @@ class val ConnectorSourceCoordinatorBuilder[In: Any val]
     host: String, service: String, cookie: String,
-    max_credits: U32, refill_credits: U32)
+    max_credits: U32, refill_credits: U32,
+    barrier_coordinator: BarrierCoordinator)
```
same from above

jtfmumm commented on examples/pony/dumb_passthrough/README.md:
```diff
@@ -0,0 +1,113 @@
+# Alphabet
```
This README is for the alphabet app

slfritchie replied:
I'm removing all passthrough apps in the next commit, after the current round of CI tests finish.

jtfmumm commented on examples/pony/dumb_passthrough/.gitignore:
```diff
@@ -0,0 +1,7 @@
+alphabet
```
shouldn't be alphabet for this app.

jtfmumm commented on examples/pony/dumb_passthrough/dumb_passthrough.pony:
```diff
@@ -0,0 +1,70 @@
+/*
```
Is there a reason we have a separate passthrough app for the alosource and TCPSource?  Elsewhere, we use command line args to choose between GenSource and TCPSource.  Why not do that with the passthrough app?

jtfmumm commented on examples/pony/multi_worker_forwarding/.gitignore:
```diff
@@ -0,0 +1,7 @@
+alphabet
```
shouldn't be alphabet for this app.

jtfmumm commented on examples/pony/multi_worker_forwarding/multi_worker_forwarding.pony:
```diff
@@ -0,0 +1,78 @@
+/*
```
Similar "why is this its own app?" question here.  What makes this different from the other two passthrough apps, and can they be consolidated?

jtfmumm commented on examples/pony/passthrough/.gitignore:
```diff
@@ -0,0 +1,7 @@
+alphabet
```
shouldn't be alphabet for this app.

jtfmumm commented on lib/wallaroo/core/sink/connector_sink/connector_sink.pony:
```diff
@@ -504,35 +556,57 @@ actor ConnectorSink is Sink
 
-      _phase = QueuingSinkPhase(_sink_id,
-        this)
+      try
```
This is inside out for how we normally do the state transitions.  Can we add a method on `SinkPhase` that triggers `Fail()` on anything but `BarrierSinkPhase`, but on `BarrierSinkPhase` causes it to call a different method on `ConnectorSink` for the queuing transition, passing it's queue along?

slfritchie replied:
Fixed by 19276f379d5cbd2ae536a14e6aed37e72452f365

slfritchie commented:
@jtfmumm: The threading of the new `barrier_coordinator` argument is so that the `ConnectorSource` has access to it so that `ConnectorSource` could abort a checkpoint.  The traits for `SourceCoordinator` and `SourceCoordinatorBuilderBuilder` meant that the extra arg needed adding to all sources.  The `_` suffix was me trying to indicate that the argument isn't used.

WRT the `aloc_passthrough` app being a separate thing ... PR #2945 is still open.  It would be easier to merge features with the `passthrough` app if #2945 were merged.

jtfmumm commented:
@slfritchie I'm ok with delaying the merge of passthrough-related apps until after we merge PR #2945  and keeping it out of the scope of this PR.

slfritchie commented:
PR #2945 and been closed and replaced with PR #3016.

slfritchie commented:
I've added the do-not-merge label and also rebased (`git rebase --onto master 460b86b^ aloc-tee-source`).  Final re-review & merge will take place after John returns from holiday.

slfritchie commented:
Hrmm, https://circleci.com/gh/WallarooLabs/wallaroo/23889?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link failed with this complaint, which is new to me and AFAICT not in an existing GH ticket:

```
1569527889.166938,Received CheckpointBarrierToken(3) when still processing AutoscaleResumeBarrierToken(initializer, 1) at step 133881423716139036102229944472886383738
This should never happen: failure in /root/project/lib/wallaroo/core/step/step_phase.pony at line 222
```

slfritchie commented:
@jtfmumm: Friendly reminder to re-review on Monday.

jtfmumm commented:
@slfritchie This looks ready to merge after (1) unused variables are removed [i.e. the ones you're suffixing with `_`] and (2) commits are squashed into logical commits.

slfritchie commented:
TODO: also remove the added printf spam.

slfritchie commented:
Squashing done by keeping commits with a 'fix' in the commit message as is, and squashing non-'fix' items as they appear in the commit log.
